### PR TITLE
Fix broken edge case of async function lowering

### PR DESCRIPTION
**Pull request by rtsao**
Fixes https://github.com/evanw/esbuild/issues/388 and adds regression test.

Also adds an optimization to skip argument forwarding altogether when it is trivial to determine that all default parameter expressions cannot throw.

### CODE REVIEW

File: internal/js_parser/js_parser_lower.go [hunk 1]
```diff
@@ -268,35 +268,26 @@ func (p *parser) lowerFunction(
 					*args = append(*args, js_ast.Arg{Binding: js_ast.Binding{Loc: arg.Binding.Loc, Data: &js_ast.BIdentifier{Ref: argRef}}})
 				}
 
-				// Forward all arguments from the outer function to the inner function
-				if p.fnOnlyDataVisit.argumentsRef != nil {
```
Comment (rtsao) [inline internal/js_parser/js_parser_lower.go:272]: The problem seems to be `argumentsRef` is non-nil if the async arrow function is nested inside a regular function. This code path only triggers when the parameters have a default value (and hence `couldThrowErrors = true`), so both conditions are required for this bug.

File: internal/js_parser/js_parser.go [hunk 2]
```diff
@@ -8987,7 +8987,7 @@ func (p *parser) visitExprInOut(expr js_ast.Expr, in exprIn) (js_ast.Expr, exprO
 		p.pushScopeForVisitPass(js_ast.ScopeFunctionBody, e.Body.Loc)
 		e.Body.Stmts = p.visitStmtsAndPrependTempRefs(e.Body.Stmts)
 		p.popScope()
-		p.lowerFunction(&e.IsAsync, &e.Args, e.Body.Loc, &e.Body.Stmts, &e.PreferExpr, &e.HasRestArg)
+		p.lowerFunction(&e.IsAsync, &e.Args, e.Body.Loc, &e.Body.Stmts, &e.PreferExpr, &e.HasRestArg, true)
```
Comment (rtsao) [inline internal/js_parser/js_parser.go:8990]: I suppose adding something like `&e.isArrow` might be more readable. But I figured omitting such a property on `Fn` was intentional?

File: scripts/end-to-end-tests.js [hunk 3]
```diff
@@ -1032,6 +1032,20 @@ in.js:24:30: warning: Writing to getter-only property "#getter" will throw
         )
       `,
     }, { async: true }),
+    test(['in.js', '--outfile=node.js', '--target=es6'],
+    {
+      'in.js': `
+        function nonArrowWrapper() {
+          return async (x, paramWithDefault = {}) => {
```
Comment (rtsao) [inline scripts/end-to-end-tests.js:1039]: Some expressions can be statically determined to never throw. Of these, it might be worth handling the subset that are both fairly common and trivial to identify (e.g. `false`, `true`, etc.)

File: internal/js_parser/js_parser.go (hunk 2 above)
Comment (evanw) [reply internal/js_parser/js_parser.go:8990]: Yes, that's correct. This change sounds good.

File: internal/js_parser/js_parser_lower.go (hunk 1 above)
Comment (evanw) [reply internal/js_parser/js_parser_lower.go:272]: Thanks for the summary.

**Review by evanw — APPROVED** [review]
This looks good. Thanks for fixing this.
